tests/pubsub/utils.nim

```diff
@@ -8,10 +8,10 @@ proc generateNodes*(num: Natural, gossip: bool = false): seq[Switch] =
     result.add(newStandardSwitch(gossip = gossip))
 
 proc subscribeNodes*(nodes: seq[Switch]) {.async.} =
-  var dials: seq[Future[void]]
+  var dials: seq[Future[Connection]]
   for dialer in nodes:
     for node in nodes:
       if dialer.peerInfo.peerId != node.peerInfo.peerId:
-        dials.add(dialer.subscribeToPeer(node.peerInfo))
+        dials.add(dialer.dial(node.peerInfo))
```

Comment:
here's the exception-unsafe code for this PR: if `dial` raises, the for loop will be interrupted with some peers dialed and some not - that's generally not safe to reason about.

dial above looks unsafe in the same way: if one of the transports fails with an exception, it will raise and prevent trying the other transports.

Reply:
@arnetheduck its actually will never raise, because `Future[T]` result is not read, so exception could not be raised.

Reply:
right. _sigh_

Reply:
dial itself though has a loop in it that will raise (while looping over transports/addrs), because it awaits..

Reply:
Of course but looks like we do not care about `dial` errors here.

Reply:
well, maybe, no idea what the intent is here, but I'm pretty sure that silently dropping all errors including defects is not part of it ;) 

I'm also pretty sure that if one transport fails in `dial`, trying another makes sense

Reply:
yes, thats a good point, I'll handle that.

Reply:
btw, this are tests, which run in a very controlled environment so I wouldn't worry about this too much, but the same problem was present in `siwtch.dial`, which is now fixed.

Reply:
code in tests gets copy-pasted - treat them the same as any other production code.

Reply:
which means that in this case, you shouldn't be collecting futures in an array without doing something to them - it's a smell. the await/sleep will introduce spurious test failures / non-determinism which will confuse other people running the test. it also slows down the test itself by being longer than what it would have been had the test been deterministic. this adds up.

Reply:
This speeds things up because it effectively dials all those nodes in parallel, this allows scaling to more peers eventually. No spurious test failures either, if one of the futures failed, it will throw and the whole test will terminate with an error, which is what you want either way.

Reply:
the slowdown is because  of the `sleepAsync(100)` below - it unconditionally sleeps regardless  if these futures have completed.
putting futures in a seq to parallelize is fine, not doing anything with the futures therein at the end is a problem / smell - as cheatfate points out, awaiting them is the right thing to do - it will work optimally on machines of all speeds

Reply:
Sorry, but how it can throw? `await allFutures()` do not raise if one of the Futures are failed. You should check all your Future[T] results after `await allFutures()`.

Reply:
we established it can't - but nonetheless, something should be done with those futures - else they just leak and hide runtime issues that happen